// vuetify/es5/components/VTextField/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _VTextField = require('./VTextField');

var _VTextField2 = _interopRequireDefault(_VTextField);

var _VTextarea = require('../VTextarea/VTextarea');

var _VTextarea2 = _interopRequireDefault(_VTextarea);

var _rebuildFunctionalSlots = require('../../util/rebuildFunctionalSlots');

var _rebuildFunctionalSlots2 = _interopRequireDefault(_rebuildFunctionalSlots);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// TODO: remove this in v2.0
var wrapper = {
    functional: true,
    props: {
        textarea: Boolean,
        multiLine: Boolean
    },
    render: function render(h, _ref) {
        var props = _ref.props,
            data = _ref.data,
            slots = _ref.slots,
            parent = _ref.parent;

        var children = (0, _rebuildFunctionalSlots2.default)(slots(), h);
        if (props.textarea) {
            console.warn('[Vuetify] \'<v-text-field textarea>\' is deprecated, use \'<v-textarea outline>\' instead in ' + parent._name);
        }
        if (props.multiLine) {
            console.warn('[Vuetify] \'<v-text-field multi-line>\' is deprecated, use \'<v-textarea>\' instead in ' + parent._name);
        }
        if (props.textarea || props.multiLine) {
            data.attrs.outline = props.textarea;
            return h(_VTextarea2.default, data, children);
        } else {
            return h(_VTextField2.default, data, children);
        }
    }
};
/* istanbul ignore next */
wrapper.install = function install(Vue) {
    Vue.component(_VTextField2.default.name, wrapper);
};
exports.default = wrapper;
